sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/functions.scala

```diff
@@ -857,3 +857,329 @@ object HyperLogLogPlusPlus {
+        m3 += -3.0 * deltaN * m2 + delta * (delta2 - deltaN2)
+        buffer.setDouble(mutableAggBufferOffset + 3, m3)
+      }
+
+      if (momentOrder >= 4) {
+        m4 = buffer.getDouble(mutableAggBufferOffset + 4)
+        m4 += -4.0 * deltaN * m3 - 6.0 * deltaN2 * m2 +
+          delta * (delta * delta2 - deltaN * deltaN2)
+        buffer.setDouble(mutableAggBufferOffset + 4, m4)
+      }
+    }
+  }
+
+  /**
+   * Merge two central moment buffers.
+   */
+  override def merge(buffer1: MutableRow, buffer2: InternalRow): Unit = {
+    val n1 = buffer1.getDouble(mutableAggBufferOffset)
+    val n2 = buffer2.getDouble(inputAggBufferOffset)
+    val mean1 = buffer1.getDouble(mutableAggBufferOffset + 1)
+    val mean2 = buffer2.getDouble(inputAggBufferOffset + 1)
+
+    var secondMoment1 = 0.0
+    var secondMoment2 = 0.0
+    var secondMoment = 0.0
+
+    var thirdMoment1 = 0.0
+    var thirdMoment2 = 0.0
+    var thirdMoment = 0.0
+
+    var fourthMoment1 = 0.0
+    var fourthMoment2 = 0.0
+    var fourthMoment = 0.0
+
+    val n = n1 + n2
+    delta = mean2 - mean1
+    deltaN = delta / n
+    val mean = mean1 + deltaN * n2
+
+    buffer1.setDouble(mutableAggBufferOffset, n)
+    buffer1.setDouble(mutableAggBufferOffset + 1, mean)
+
+    // higher order moments computed according to:
+    // https://en.wikipedia.org/wiki/Algorithms_for_calculating_variance#Higher-order_statistics
+    if (momentOrder >= 2) {
+      secondMoment1 = buffer1.getDouble(mutableAggBufferOffset + 2)
+      secondMoment2 = buffer2.getDouble(inputAggBufferOffset + 2)
+      secondMoment = secondMoment1 + secondMoment2 + delta * deltaN * n1 * n2
+      buffer1.setDouble(mutableAggBufferOffset + 2, secondMoment)
+    }
+
+
+    if (momentOrder >= 3) {
+      thirdMoment1 = buffer1.getDouble(mutableAggBufferOffset + 3)
+      thirdMoment2 = buffer2.getDouble(inputAggBufferOffset + 3)
+      thirdMoment = thirdMoment1 + thirdMoment2 + deltaN * deltaN * delta * n1 * n2 *
+        (n1 - n2) + 3.0 * deltaN * (n1 * secondMoment2 - n2 * secondMoment1)
+      buffer1.setDouble(mutableAggBufferOffset + 3, thirdMoment)
+    }
+
+    if (momentOrder >= 4) {
+      fourthMoment1 = buffer1.getDouble(mutableAggBufferOffset + 4)
+      fourthMoment2 = buffer2.getDouble(inputAggBufferOffset + 4)
+      fourthMoment = fourthMoment1 + fourthMoment2 + deltaN * deltaN * deltaN * delta * n1 *
+        n2 * (n1 * n1 - n1 * n2 + n2 * n2) + deltaN * deltaN * 6.0 *
+        (n1 * n1 * secondMoment2 + n2 * n2 * secondMoment1) +
+        4.0 * deltaN * (n1 * thirdMoment2 - n2 * thirdMoment1)
+      buffer1.setDouble(mutableAggBufferOffset + 4, fourthMoment)
+    }
+  }
+
+  /**
+   * Compute aggregate statistic from sufficient moments.
+   */
+  def getStatistic(n: Double, moments: Array[Double]): Double
+
+  override final def eval(buffer: InternalRow): Any = {
+    val n = buffer.getDouble(mutableAggBufferOffset)
+    val moments = sufficientMoments.map { momentIdx =>
```

Comment:
Per Xiangrui's comments about passing buffer to subclasses, I think it's necessary to do it this way. We don't have to pass the InternalRow object to the subclasses, and just pass the sufficient moments instead. See below discussion for more details.